ddtrace/internal/remoteconfig/_pubsub.py

```diff
@@ -0,0 +1,114 @@
+register the same instance of PubSub class on different products, we would have one thread waiting to the main process.
+
+Each DD Product (APM, ASM, DI, CI) may implement its PubSub Class.
+
+Example 1: A callback for one or more Remote Configuration Products
+-------------------------------------------------------------------
+AppSec needs to aggregate different products in the same callback for all child processes.
+
+class AppSecRC(PubSubMergeFirst):
+    __shared_data = ConnectorSharedMemory()
+
+    def __init__(self, _preprocess_results, callback, name="Default"):
+        self._publisher = self.__publisher_class__(self.__shared_data, _preprocess_results)
+        self._subscriber = self.__subscriber_class__(self.__shared_data, callback, name)
+
+asm_callback = AppSecRC(preprocess_1click_activation, appsec_callback, "ASM")
+
+remoteconfig_poller.register("ASM_PRODUCT", asm_callback)
+remoteconfig_poller.register("ASM_FEATURES_PRODUCT", asm_callback)
+
+
+Example 2: One Callback for each product
+----------------------------------------
+DI needs to aggregate different products in the same callback for all child processes.
+
+class DynamicInstrumentationRC(PubSub):
+    __shared_data = ConnectorSharedMemory()
+
+    def __init__(self, _preprocess_results, callback, name="Default"):
+        self._publisher = self.__publisher_class__(self.__shared_data, _preprocess_results)
+        self._subscriber = self.__subscriber_class__(self.__shared_data, callback, name)
+
+di_callback_1 = DynamicInstrumentationRC(callback=di_callback_1, name="ASM")
+di_callback_2 = DynamicInstrumentationRC(callback=di_callback_2, name="ASM")
+
+remoteconfig_poller.register("DI_1_PRODUCT", di_callback)
+remoteconfig_poller.register("DI_2_PRODUCT", di_callback_2)
+
+"""
+import abc
+
+import six
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.remoteconfig._connectors import ConnectorBase
+from ddtrace.internal.remoteconfig._publishers import RemoteConfigPublisher
+from ddtrace.internal.remoteconfig._publishers import RemoteConfigPublisherBase
+from ddtrace.internal.remoteconfig._publishers import RemoteConfigPublisherMergeFirst
+from ddtrace.internal.remoteconfig._subscribers import RemoteConfigSubscriber
+
+
+log = get_logger(__name__)
+
+
+class PubSubBase(six.with_metaclass(abc.ABCMeta)):
+    __publisher_class__ = RemoteConfigPublisherBase
+    __subscriber_class__ = RemoteConfigSubscriber
+    __shared_data = None  # type: ConnectorBase
+    _publisher = None  # type: RemoteConfigPublisherBase
+    _subscriber = None  # type: RemoteConfigSubscriber
+
+    def start_subscriber(self):
+        self._subscriber.start()
+
+    def restart_subscriber(self):
+        self._subscriber.force_restart()
+
+    def _poll_data(self, test_tracer=None):
+        self._subscriber._get_data_from_connector_and_exec(test_tracer=test_tracer)
+
+    def stop(self):
+        self._subscriber.stop()
+
+
+class PubSub(PubSubBase):
```

Comment:
I wonder if we could simplify this design a little bit. We already have the publisher, the subscriber and the event bus (the connector), so we could potentially do away without this class. I would tend to see the product-specific subclasses more as simply sub-classes of a subscriber than this pub-sub hub. WDYT?

Reply:
Yeah, it's true, we can simplify this file to use just one class but we need at least one which contains the abstract methods and common methods of the pubsub class and each product-specific implement/inherits from that class, right?

Reply:
yep